### ISSUE TITLE
[BUG] cachedb_redis: spaces in Redis arguments do not work, probably due to lack of command interpolation support

### ISSUE THREAD

**Issue by olofhe** [OpenSIPS/opensips#2036]
Hi!

The `cachedb_redis` module does not seem to support raw Redis commands where arguments contain spaces. Attempting to use such arguments results in errors, that indicate that the command string is not being correctly parsed.

E.g., this command

```
cache_raw_query("redis", "SET foo \"bar baz\"", "$avp(result)");
```

throws:

```
ERROR:cachedb_redis:redis_raw_query_send: Redis operation failure - ERR syntax error
```

and this

```
cache_raw_query("redis", "EVAL \"return 'foo'\" 0", "$avp(result)")
```

throws: 

```
ERROR:cachedb_redis:redis_raw_query_send: Redis operation failure - ERR value is not an integer or out of range
```

As you can see, this issue not only affects the possibility to use spaces in keys and values. It also makes usage of the `EVAL` command impossible, since the Lua script passed as an argument to Redis naturally requires spaces.

I think the problem is that OpenSIPS does not use the command interpolation feature of Hiredis. Using the C Hiredis library I can replicate the above behavior using the corresponding commands without interpolation, getting the exact same error messages.

E.g. in C, the following works (interpolation):

```c
redisCommand(context, "SET foo %s", "bar baz")
```

but the following does not work (no interpolation):

```c
redisCommand(context, "SET foo \"bar baz\"")
```

Best Regards,
Olof

**Comment by zenichev** [comment]
I described this issue here:
http://lists.opensips.org/pipermail/users/2020-March/042590.html

From what I understand an implementation of this command run, is here:
https://github.com/OpenSIPS/opensips/blob/9578c7fef7f54000b9e38c5b248302649a739d06/modules/cachedb_redis/cachedb_redis_dbase.c#L281

Also if I understand this correctly, there is no string formatting implemented in:
`redisCommand(node->context,fmt,##args);`

not quite sure what "##args" makes here though.
But "fmt" seems to be formatted in advance, before the run of redisCommand().

**Comment by olofhe** [comment]
I forgot to include the issue that @zenichev referenced: https://github.com/OpenSIPS/opensips/issues/1250 . See, in particular https://github.com/OpenSIPS/opensips/issues/1250#issuecomment-510065176, that states that the input string is simply passed to the Hiredis `redisvCommand` function. My above examples used the `redisCommand` function instead, but I assume they have the same behavior in this regard.

**Comment by liviuchircu** [comment]
Hey @olofhe and @zenichev!  Sorry that this issue took so long, but the issue should be definitively fixed on `master` branch, per the above commit and examples.

Still, due to the importance of the fix, I've also backported it to `3.1` and `3.2`, but it's disabled by default.  In order to make use of these quoted strings in your raw queries, you will have to enable this modparam first:

    modparam("cachedb_redis", "enable_query_arg_quoting", 1)

**Comment by liviuchircu** [comment]
Yes, @olofhe!  But you will have to use both quote types at the same time in order to achieve that.  For example, if you want to store a JSON with lots of double-quotes into Redis, try to enclose it in single-quotes:

```
cache_raw_query("redis", "SET foo '$json_compact(my_json)'", "$avp(result)");
```
